tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -11352,6 +11352,111 @@ bool Sema::DiagnoseHLSLMethodCall(const CXXMethodDecl *MD, SourceLocation Loc) {
   return false;
 }
 
+
+void ValidateCallGraphWaveSize(clang::Sema *S, FunctionDecl *FD) {
+  const auto *SM =
+      hlsl::ShaderModel::GetByName(S->getLangOpts().HLSLProfile.c_str());
+
+  const std::string &entryName = S->getLangOpts().HLSLEntryFunction;
+  StringRef functionName = "";
+  if (FD->getIdentifier())
+    functionName = FD->getIdentifier()->getName();
+  bool isEntry = false;
+  bool isNode = false;
+  bool isCS = false;
+
+  isEntry = !SM->IsLib() &&
+            FD->getDeclContext()->getDeclKind() == Decl::Kind::TranslationUnit &&
+            functionName == entryName;
+
+  for (auto *pAL : FD->specific_attrs<HLSLShaderAttr>()) {
+    StringRef Literal = pAL->getStage();
+    DXIL::ShaderKind Stage =
+        ShaderModel::KindFromFullName(Literal);
+    isNode |= Stage == DXIL::ShaderKind::Node;
+    isCS |= Stage == DXIL::ShaderKind::Compute;
+  }
+
+  HLSLWaveSizeAttr *attr = FD->getAttr<HLSLWaveSizeAttr>();
+  
+  if (!SM->IsLib() && !isEntry && !isNode) {
+    S->Diag(attr->getRange().getBegin(),
+            diag::err_hlsl_attribute_valid_on_entry_function_only)
```

Comment:
This is going to be a problematic regression. We currently allow inactive entry points in shaders like:

```hlsl
[WaveSize(32)]
[numthreads(32, 1, 1)]
void csmain1(){ }

[WaveSize(32)]
[numthreads(32, 1, 1)]
void csmain2(){ }
```
[Compiler Explorer Link](https://godbolt.org/z/3qMMssM39)

And entry points can be called as regular functions, so the call graph walking won't help.

Reply:
I've added a test with this code (WaveSizeAttrAllowed4.hlsl), my changes don't regress successful compilation of the code you've posted above.
For extra clarity, I should note that the code inside DiagnoseTranslationUnit() is specifically for validation of entry point functions. There is separate code in DiagnoseHLSLDecl that is responsible for diagnosing lib shaders that are *not* entry points. The code locations are necessary because those are the only locations that validate the declarations that are of interest at that point. By that I mean, DiagnoseTranslationUnit() is the only convenient location that has a natural way to validate all functions reachable from the entry point. This location is worthless for shaders that are targeting libraries on compilation. Conversely, the DiagnoseHLSLDecl() location will be executed on all function declarations if the shader is targeting a library profile, which changes the rules of validation.